easy_exceptions/easy_exceptions.py

```diff
@@ -17,13 +32,30 @@ def EasyException(name=None, parent=None):
     if not name.isalnum():
         raise ValueError("Exception name {} is not alphanumeric".format(name))
 
-    if not parent:
-        parent = bound_parent
-        
+    # perform a check outside of a lock to allow full parallelism when exceptions have already been created
     exc_class = generated_exceptions.get((name, parent))
     if exc_class:
-        return exc_class 
-    else:
-        exec("class {}(parent):\n    pass".format(name)) 
-        generated_exceptions[(name, parent)] = locals()[name]
-        return generated_exceptions[(name, parent)]
+        return exc_class
+
+    acquired = update_mutex.acquire()
+    if not acquired:
+        raise Exception("Lock acquire while generating exception failed")
```

Comment:
@Cruuncher same thing as above, just a suggestion for a more easy to understand exception message.